iphone/Maps/Core/iCloud/SynchronizationStateResolver.swift

```diff
@@ -0,0 +1,224 @@
+    LOG(.info, "Local contents: \(currentLocalContents.count)")
+    currentLocalContents.sorted(by: { $0.fileName < $1.fileName }).forEach { LOG(.info, $0.shortDebugDescription) }
+    LOG(.info, "Events to process: \(outgoingEvents.count)")
+    outgoingEvents.forEach { LOG(.info, $0) }
+
+    return outgoingEvents
+  }
+
+  func resetState() {
+    LOG(.debug, "Resetting state")
+    currentLocalContents.removeAll()
+    currentCloudContents.removeAll()
+    localContentsGatheringIsFinished = false
+    cloudContentGatheringIsFinished = false
+  }
+
+  func setInitialSynchronization(_ isInitialSynchronization: Bool) {
+    self.isInitialSynchronization = isInitialSynchronization
+  }
+
+  private func resolveDidFinishGathering(localContents: LocalContents, cloudContents: CloudContents) -> [OutgoingSynchronizationEvent] {
+    currentLocalContents = localContents
+    currentCloudContents = cloudContents
+    guard localContentsGatheringIsFinished, cloudContentGatheringIsFinished else { return [] }
+    
+    switch (localContents.isEmpty, cloudContents.isEmpty) {
+    case (true, true):
+      return []
+    case (true, false):
+      return cloudContents.map { .createLocalItem(with: $0) }
+    case (false, true):
+      return localContents.map { .createCloudItem(with: $0) }
+    case (false, false):
+      var events = [OutgoingSynchronizationEvent]()
+      if isInitialSynchronization {
+        /* During the initial synchronization:
+         - all conflicted local and cloud items will be saved to avoid a data loss
+         - all items that are in the cloud but not in the local container will be created in the local container
+         - all items that are in the local container but not in the cloud container will be created in the cloud container
+         */
+        localContents.forEach { localItem in
+          if let cloudItem = cloudContents.firstByName(localItem), localItem.lastModificationDate != cloudItem.lastModificationDate {
+            events.append(.resolveInitialSynchronizationConflict(localItem))
+            events.append(.updateLocalItem(with: cloudItem))
+          }
+        }
+
+        let itemsToCreateInCloudContainer = localContents.filter { !cloudContents.containsByName($0) }
+        let itemsToCreateInLocalContainer = cloudContents.filter { !localContents.containsByName($0) }
+        events.append(contentsOf: itemsToCreateInCloudContainer.map { .createCloudItem(with: $0) })
+        events.append(contentsOf: itemsToCreateInLocalContainer.map { .createLocalItem(with: $0) })
+
+        events.append(.didFinishInitialSynchronization)
+        isInitialSynchronization = false
+        return events
+      } else {
+        var events = [OutgoingSynchronizationEvent]()
+
+        cloudContents.getErrors.forEach { events.append(.didReceiveError($0)) }
+        cloudContents.withUnresolvedConflicts.forEach { events.append(.resolveVersionsConflict($0)) }
+
```

Comment:
Add a comment that ```itemsToCreateInCloudContainer``` is not present here because ...

If a user turned-off iCloud, created some items and then turned-on, ```isInitialSynchronization``` branch works?

Reply:
> Add a comment that `itemsToCreateInCloudContainer` is not present here because ...

done!

> If a user turned-off iCloud, created some items and then turned-on, isInitialSynchronization branch works?
Yes. The initial sync will be triggered.

In this case, items with the same name but different content will be preserved. This issue can be fixed by implementing smart bookmarks merging. But it's still hard to find out what bookmarks the user wants to save and what to remove until we don't make a journal of actions...